provider/datagen/src/driver.rs

```diff
@@ -5,20 +5,265 @@
+            Some(RuntimeFallbackLocation::External) => "with external fallback",
+        };
+        write!(f, "{start}, ")?;
+        match self.deduplication_strategy {
+            Some(x) => x.describe(f)?,
+            None => write!(f, "default deduplication")?,
+        }
+        write!(f, ", and these locales: {:?}", sorted_locales)
+    }
+}
+
+/// Wraps the language identifiers with [`LocaleWithExpansion::with_variants`]
+impl FromIterator<LanguageIdentifier> for LocalesWithFallback {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = LanguageIdentifier>>(iter: T) -> Self {
+        Self::from_iter(iter.into_iter().map(LocaleWithExpansion::with_variants))
+    }
+}
+
+impl FromIterator<LocaleWithExpansion> for LocalesWithFallback {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = LocaleWithExpansion>>(iter: T) -> Self {
+        Self {
+            runtime_fallback_location: Default::default(),
+            deduplication_strategy: Default::default(),
+            locales: iter.into_iter().collect(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum LocalesWithOrWithoutFallback {
+    WithFallback(LocalesWithFallback),
+    WithoutFallback(LocalesWithoutFallback),
+}
+
+impl LocalesWithOrWithoutFallback {
+    /// Converts this field to [`LocalesWithoutFallback`] and returns a reference to it.
+    pub(crate) fn coerce_without_fallback(&mut self) -> &mut LocalesWithoutFallback {
+        match self {
+            Self::WithoutFallback(config) => config,
+            Self::WithFallback(config) => {
+                let locales_set = core::mem::take(&mut config.locales);
+                let config = locales_set.into_iter().map(|x| x.into_langid()).collect();
+                *self = Self::WithoutFallback(config);
+                let Self::WithoutFallback(config) = self else {
+                    unreachable!()
+                };
+                config
+            }
+        }
+    }
+
+    /// Converts this field to [`LocalesWithFallback`] and returns a reference to it.
+    pub(crate) fn coerce_with_fallback(&mut self) -> &mut LocalesWithFallback {
+        match self {
+            Self::WithFallback(config) => config,
+            Self::WithoutFallback(config) => {
+                let locales_set = core::mem::take(&mut config.locales);
+                let config = locales_set.into_iter().collect();
+                *self = Self::WithFallback(config);
+                let Self::WithFallback(config) = self else {
+                    unreachable!()
+                };
+                config
+            }
+        }
+    }
+
+    /// Returns whether this is exactly the all-locales value.
+    pub(crate) fn is_all_locales(&self) -> bool {
+        let Self::WithFallback(config) = self else {
+            return false;
+        };
+        let mut it = config.locales.iter();
+        matches!((it.next(), it.next()), (Some(lid), None) if lid == &LocaleWithExpansion::wildcard())
```

Comment:
if you make `wildcard()` an associated `const` you can match against it.

Reply:
For now I'd rather keep it a function because it's more flexible.